File: 2024/day25.py
import AOCInit
import util

if __name__ != '__main__':
    exit()

inp = util.getInput(d=25, y=2024)

pins = tuple(
        pin.splitlines()
        for pin in inp.split('\n\n')
)

locks: list[tuple[int, ...]] = []
keys: list[tuple[int, ...]] = []

for pin in pins:
    (locks if pin[0][0] == '#' else keys)\
            .append(tuple(r.count('#') - 1 for r in util.transpose(pin)))

# part 1
print(sum(
    all(a + b <= 5 for a, b in zip(k, l))
    for k in keys
    for l in locks
))


# part 2
# no part 2

File: 2024/day25.cpp
#include <vector>

#include "../util.hpp"

int main(int, char**){
    std::vector<std::vector<int>> locks, keys;
    for (const std::string &line : util::split(util::getInput(2024, 25), "\n\n")){
        std::vector<std::string> pins = util::splitline(line);
        int pinMaxLen = pins.size(), pinCount = pins[0].size();
        std::vector<int> counts;
        for (int j = 0; j < pinCount; ++j){
            int count = -1;
            for (int i = 0; i < pinMaxLen; ++i){
                count += pins[i][j] == '#';
            }
            counts.push_back(count);
        }
        if (pins[0][0] == '#'){
            locks.push_back(counts);
        } else {
            keys.push_back(counts);
        }
    }

    // part 1
    int count = 0;
    for (auto k : keys){
        for (auto l : locks){
            int n = k.size();
            bool fits = true;
            for (int i = 0; i < n; ++i){
                if (k[i] + l[i] > 5){
                    fits = false;
                    break;
                }
            }
            count += fits;
        }
    }
    util::output(count);


    // part 2
    // no part 2

    return 0;
}
